tsinfer/inference.py

```diff
@@ -1924,6 +1912,156 @@ def get_mutations(self, node_id):
+            index = np.where(times == t)[0]
+            k = index.shape[0]
+            times[index] += np.arange(k)[::-1] / k
+        tables.nodes.time = times
+
+    has_virtual_like_root = detach_node_over_grand_root(tables)
+    if has_virtual_like_root and not keep_single_mrca:
+        split_genomewide_mrca(tables)
+
+    logger.info(
+        "Sorting and simplifying with filter_sites=False, filter_populations=False, "
+        "filter_individuals=False, and keep_unary=True on "
+        f"{tables.nodes.num_rows} nodes and {ts.num_edges} edges"
+    )
+    tables.sort()
+    tables.simplify(
+        samples=sample_order,
+        filter_sites=False,
+        filter_populations=False,
+        filter_individuals=False,
+        keep_unary=True,
+    )
+    logger.info(
+        "Finished simplify; now have {} nodes and {} edges".format(
+            tables.nodes.num_rows, tables.edges.num_rows
+        )
+    )
+    return tables.tree_sequence()
+
+
+def detach_node_over_grand_root(tables):
+    # If parent is a genome-wide ancestor whose child is the grand MRCA, and no
+    # other edges point to the parent, detach the parent from the tree sequence (as
+    # it represents a virtual-root-like ancestor, added to make the LS matching work)
+    tables.sort()
+    last_edge = tables.edges[tables.edges.num_rows - 1]
+    oldest_node_id = last_edge.parent
+    if np.sum(tables.edges.parent == oldest_node_id) > 1:
+        return False  # other edges point to the oldest node => not a virtual-like root
+    if last_edge.left != 0 or last_edge.right != tables.sequence_length:
+        return False  # edge does not span the entire genome
+
+    logger.info("Removing the oldest edge to detach the virtual-root-like ancestor")
+    tables.edges.truncate(tables.edges.num_rows - 1)
+    # move any mutations above the virtual-root-like ancestor to above the grand
+    # MRCA instead (these will be mutations placed by parsimony)
+    mutations_node = tables.mutations.node
+    mutations_node[mutations_node == last_edge.parent] = last_edge.child
+    tables.mutations.node = mutations_node
+    return True
+
+
+def split_genomewide_mrca(tables):
+    # If a single oldest node is a root across the entire genome, split it up into
+    # multiple nodes all with the same node.time, whenever the node children change
+    tables.sort()
+    ts = tables.tree_sequence()
+    # Check that we do have a single root everywhere - there's probably a more
+    # efficient way to do this using the quintuply linked data structure
+    roots = {root for tree in ts.trees() for root in tree.roots}
+    if len(roots) > 1:
+        return False
+
+    logger.info("Splitting the all zeros grand MRCA into separate nodes")
+    genomewide_mrca_id = roots.pop()
+    assert genomewide_mrca_id == tables.edges[tables.edges.num_rows - 1].parent
+    genomewide_mrca = tables.nodes[genomewide_mrca_id]
+    edges_to_remove = tables.edges.parent == genomewide_mrca_id
```

Comment:
It's neater to reverse the `ts.edges()` iterator, I think, so that we can use the IDs (and if we do it on the ts rather than the tables, we are assured that the order is correct):

```python
    root_breaks = set()
    keep_edges = 0
    for e in reversed(ts.edges()):
        # all the edges at the end are the ones to the genomewide_mrca
        if e.parent == genomewide_mrca_id:
            root_breaks |= {e.left, e.right}
        else:
            keep_edges = e.id + 1
            break
    tables.edges.truncate(keep_edges)
```

Reply:
I'd find the while-loop version easier (and shorter) to read tbh, but :shrug:

Reply:
There's a bug in that one: I think we want to `truncate(j+1)`, which was another reason for spelling out `keep_edges` explicitly. But happy to switch back.